src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampRequest.cs

```diff
@@ -0,0 +1,484 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Pkcs.Asn1;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    public sealed class Rfc3161TimestampRequest
+    {
+        private byte[] _encodedBytes;
+        private Rfc3161TimeStampReq _parsedData;
+
+        private Rfc3161TimestampRequest()
+        {
+        }
+
+        public int Version => _parsedData.Version;
+        public ReadOnlyMemory<byte> GetMessageHash() => _parsedData.MessageImprint.HashedMessage;
+        public Oid HashAlgorithmId => _parsedData.MessageImprint.HashAlgorithm.Algorithm;
+        public Oid RequestedPolicyId => _parsedData.ReqPolicy;
+        public bool RequestSignerCertificate => _parsedData.CertReq;
+        public ReadOnlyMemory<byte>? GetNonce() => _parsedData.Nonce;
+        public bool HasExtensions => _parsedData.Extensions?.Length > 0;
+
+        public X509ExtensionCollection GetExtensions()
+        {
+            var coll = new X509ExtensionCollection();
+
+            if (!HasExtensions)
+            {
+                return coll;
+            }
+
+            X509ExtensionAsn[] rawExtensions = _parsedData.Extensions;
+
+            foreach (X509ExtensionAsn rawExtension in rawExtensions)
+            {
+                X509Extension extension = new X509Extension(
+                    rawExtension.ExtnId,
+                    rawExtension.ExtnValue.ToArray(),
+                    rawExtension.Critical);
+
+                // Currently there are no extensions defined.
+                // Should this dip into CryptoConfig or other extensible
+                // mechanisms for the CopyTo rich type uplift?
+                coll.Add(extension);
+            }
+
+            return coll;
+        }
+
+        public async Task<Rfc3161TimestampToken> SubmitRequestAsync(Uri uri, TimeSpan timeout)
+        {
+            if (uri == null)
+                throw new ArgumentNullException(nameof(uri));
+            if (!uri.IsAbsoluteUri)
+                throw new ArgumentOutOfRangeException(nameof(uri), SR.Cryptography_TimestampReq_HttpOrHttps);
+            if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
+                throw new ArgumentOutOfRangeException(nameof(uri), SR.Cryptography_TimestampReq_HttpOrHttps);
+
+            byte[] responseContents;
+            HttpClient httpClient = null;
+
+            try
+            {
+                httpClient = new HttpClient
```

Comment:
Is a caller ever going to have to customize this `HttpClient` instance somehow, perhaps by adding proxy info or extra request headers? Does it make sense to accept a custom `HttpClientHandler` or similar as an optional parameter?

Full disclosure: I worked on a timestamping server in the past which required an _Authorization_ header before it would grant access. You can also imagine a timestamping server which requires client cert auth to prevent arbitrary callers from being able to take server resources.

Reply:
Are you thinking an overload, or replace timeout with httpClient in the signature?

Reply:
An overload, probably? TBH I'm fine with adding this overload in the future if the need arises. We should just be mindful that we may receive requests for this capability.